File: 621. Task Scheduler/solution.py
class Solution:
    def leastInterval(self, tasks, n):
        """
        :type tasks: List[str]
        :type n: int
        :rtype: int
        """
        # calculate frequency
        map = {}
        for t in tasks:
            if t in map:
                map[t] += 1
            else:
                map[t] = 1
                
        # create heap for freq
        heap = []
        for t, f in map.items():
            heapq.heappush(heap, -f)
            
        aux = []
        count = 0
        while heap:
            i = 0
            while i <= n and heap:
                f = heapq.heappop(heap)
                f += 1
                if f:
                    aux.append(f)
                i += 1
                
            if aux:
                count += n+1
            else:
                count += i
                
            while aux:
                heapq.heappush(heap, aux.pop())
            
        
        return count

File: 621. Task Scheduler/solution.java
class Solution {
    public int leastInterval(char[] tasks, int n) {
        // construct map
        Map<Character, Integer> map = new HashMap<Character, Integer>();
        for (char task: tasks) {
            if (map.containsKey(task)) {
                map.put(task, map.get(task) + 1);
            } else {
                map.put(task, 1);
            }
        }
        
        // create heap
        PriorityQueue<Task> heap = new PriorityQueue<Task>(map.size(), (a, b) -> b.priority - a.priority);
        Task task;
        for (char t: map.keySet()) {
            task = new Task(t, map.get(t));
            heap.add(task);
        }
        
        // extract heap till empty
        int intervals = 0;
        List<Task> aux = new ArrayList<Task>(map.size());
        while (!heap.isEmpty()) {
            int i = n;
            while (!heap.isEmpty() && i >= 0) {
                task = heap.poll();
                task.priority--;
                if (task.priority > 0) {
                    aux.add(task);
                }
                intervals++;
                i--;
            }
            
            if (i >= 0 && !aux.isEmpty())
                intervals += i + 1;
            
            heap.addAll(aux);
            aux.clear();
        }
        return intervals;
    }
    
    class Task {
        private Integer priority;
        private Character id;
        
        public Task(Character t, Integer p) {
            id = t;
            priority = p;
        }
    }
}
